File: sdk/python/src/watcher_sdk/usecases/capture_event.py
# Use case: validate caller input and put one event into the buffer.
# Runs on the application thread — must be fast (no I/O).
from watcher_sdk.buffer import EventBuffer
from watcher_sdk.domain.event import EventInput

_VALID_EVENT_TYPES = {"audit", "log", "trace", "metric", "event", "security", "ai", "system", "infrastructure"}
_VALID_SEVERITIES = {"debug", "info", "warn", "error", "critical"}


class CaptureEventUseCase:
    def __init__(self, buffer: EventBuffer) -> None:
        self._buffer = buffer

    def execute(self, event: EventInput) -> None:
        _validate(event)
        self._buffer.push(event)


def _validate(event: EventInput) -> None:
    if not event.message:
        raise ValueError("message is required")
    if event.event_type not in _VALID_EVENT_TYPES:
        raise ValueError(f"invalid event_type '{event.event_type}'; valid: {_VALID_EVENT_TYPES}")
    if event.severity not in _VALID_SEVERITIES:
        raise ValueError(f"invalid severity '{event.severity}'; valid: {_VALID_SEVERITIES}")

File: sdk/js/packages/core/src/usecases/capture-event.ts
/**
 * Validates caller input and pushes one event onto the buffer.
 * Runs synchronously on the caller's thread — no I/O allowed here.
 */
import type { EventBuffer } from "../buffer.js";
import type { EventInput } from "../domain/event.js";

const VALID_EVENT_TYPES = new Set([
  "audit", "log", "trace", "metric", "event",
  "security", "ai", "system", "infrastructure",
]);

const VALID_SEVERITIES = new Set([
  "debug", "info", "warn", "error", "critical",
]);

export class CaptureEventUseCase {
  constructor(private readonly buffer: EventBuffer) {}

  execute(event: EventInput): void {
    validate(event);
    this.buffer.push(event);
  }
}

function validate(event: EventInput): void {
  if (!event.message) throw new Error("message is required");
  if (!VALID_EVENT_TYPES.has(event.eventType)) {
    throw new Error(`invalid eventType "${event.eventType}"; valid: ${[...VALID_EVENT_TYPES].join(", ")}`);
  }
  if (!VALID_SEVERITIES.has(event.severity)) {
    throw new Error(`invalid severity "${event.severity}"; valid: ${[...VALID_SEVERITIES].join(", ")}`);
  }
}
